proposals/bulk-memory-operations/Overview.md

```diff
@@ -171,93 +169,111 @@ Since WebAssembly currently does not allow for multiple memories, the memory
+| 3    | Declared, externval              |             |                 | `extern_kind` | `varuint32` | `idx`*       |
+| 4    | Legacy active, funcref elemexpr  |             | `init_expr`     |               | `varuint32` | `elem_expr`* |
+| 5    | Passive, elemexpr                |             |                 | `elem_type`   | `varuint32` | `elem_expr`* |
+| 6    | Active, elemexpr                 | `varuint32` | `init_expr`     | `elem_type`   | `varuint32` | `elem_expr`* |
+| 7    | Declared, elemexpr               |             |                 | `elem_type`   | `varuint32` | `elem_expr`* |
+
+All other flag values are illegal.  Note that the "declared" attribute
+is not used by this proposal, but is used by the reference-types
+proposal.
+
+The `extern_kind` must be zero, signifying a function definition.  An `idx` is a
+`varuint32` that references an entity in the module, currently only its function
+table.
+
+At present the table index must be zero, but the reference-types
+proposal introduces a notion of multiple tables.
 
 An `elem_expr` is like an `init_expr`, but can only contain expressions of the following sequences:
 
-| Binary | Text | Description |
-| - | - | - |
-| `0xd0 0x0b` | `ref.null end` | Returns a null reference |
+| Binary                | Text                    | Description                                |
+| -                     | -                       | -                                          |
+| `0xd0 0x0b`           | `ref.null end`          | Returns a null reference                   |
 | `0xd2 varuint32 0x0b` | `ref.func $funcidx end` | Returns a reference to function `$funcidx` |
 
-TODO: coordinate with other proposals to determine the binary encoding for `ref.null` and `ref.func`.
-
 ### Segment Initialization
 
 In the MVP, segments are initialized during module instantiation. If any segment
 would be initialized out-of-bounds, then the memory or table instance is not
 modified.
 
-This behavior is changed in the bulk memory proposal.
-
-Each active segment is initialized in module-definition order. For each
-segment, each byte in the data segment is copied into the memory, in order of
-lowest to highest addresses. If, for a given byte, the copy is out-of-bounds,
-instantiation fails and no further bytes in this segment nor further segments
-are copied. Bytes written before this point stay written.
+This behavior is changed in the bulk memory proposal:
 
-The behavior of element segment initialization is changed similarly, with the
-difference that elements are copied from element segments into tables, instead
-of bytes being copied from data segments into memories.
+Each active segment is initialized in module-definition order. For
+each segment, if reading the source or writing the destination would
+go out of bounds, then instantiation fails at that point.  Data that
+had already been written for previous (in-bounds) segments stays
+written.
```

Comment:
While for all other instructions the bounds checks are done before copying/initialization and partial copying does not occur, does partial copying occur for active segment initialization?

Reply:
Each segment is bounds-checked and initialized individually, in-order. If a segment fails a bounds check, previous segments will stay written, but the oob segment will not do any partial write, and the rest of the segments will not be written (immediate trap).

Initialization can be thought of as iteratively doing init -> drop on the active segment declarations. The bounds checking comes precisely from the behaviour of init. In the MVP, by contrast, all segments are bounds-checked before any are written.

Reply:
Oh, right, partial write does not occur _within_ a segment. Thanks.